pydantic/main.py

```diff
@@ -890,53 +907,62 @@ def __getattr__(self, item: str) -> Any:
+                    return None  # Can not return memoized handler with possibly freeform attr names
 
-            self._check_frozen(name, value)
+            cls._check_frozen(name, value)
 
-            attr = getattr(self.__class__, name, None)
+            attr = getattr(cls, name, None)
             # NOTE: We currently special case properties and `cached_property`, but we might need
             # to generalize this to all data/non-data descriptors at some point. For non-data descriptors
             # (such as `cached_property`), it isn't obvious though. `cached_property` caches the value
             # to the instance's `__dict__`, but other non-data descriptors might do things differently.
             if isinstance(attr, property):
-                attr.__set__(self, value)
+                return lambda model, _name, val: attr.__set__(model, val)
             elif isinstance(attr, cached_property):
-                self.__dict__[name] = value
-            elif self.model_config.get('validate_assignment', None):
-                self.__pydantic_validator__.validate_assignment(self, name, value)
-            elif self.model_config.get('extra') != 'allow' and name not in self.__pydantic_fields__:
-                # TODO - matching error
-                raise ValueError(f'"{self.__class__.__name__}" object has no field "{name}"')
-            elif self.model_config.get('extra') == 'allow' and name not in self.__pydantic_fields__:
-                if self.model_extra and name in self.model_extra:
-                    self.__pydantic_extra__[name] = value  # type: ignore
+                return SIMPLE_SETATTR_HANDLERS['cached_property']
+            elif cls.model_config.get('validate_assignment'):
+                return SIMPLE_SETATTR_HANDLERS['validate_assignment']
+            elif name not in cls.__pydantic_fields__:
+                if cls.model_config.get('extra') != 'allow':
+                    # TODO - matching error
+                    raise ValueError(f'"{cls.__name__}" object has no field "{name}"')
+                elif attr is None:
+                    # attribute does not exist, so put it in extra
+                    self.__pydantic_extra__[name] = value
+                    return None  # Can not return memoized handler with possibly freeform attr names
                 else:
-                    try:
-                        getattr(self, name)
-                    except AttributeError:
-                        # attribute does not already exist on instance, so put it in extra
-                        self.__pydantic_extra__[name] = value  # type: ignore
-                    else:
-                        # attribute _does_ already exist on instance, and was not in extra, so update it
-                        _object_setattr(self, name, value)
+                    # attribute _does_ exist, and was not in extra, so update it
+                    return SIMPLE_SETATTR_HANDLERS['extra_known']
```

Comment:
I'm a bit confused how we got this from the previous logic - don't we need to check `name in self.model_extra`?

Reply:
The check was redundant, as we know the extra attribute needs to go to the special `__pydantic_extra__` when model class does not define the private attribute. There is no need to check `model_extra` (which just checks `__pydantic_extra__`)